Feat/checkpoint modes

javarno opened the pull request:
Fixes [spring-projects/spring-integration-aws#108](https://github.com/spring-projects/spring-integration-aws/issues/108)

All existing checkpoint modes are now supported by the `KclMessageDrivenChannelAdapter`.
The new existing mode (periodic) works for both `KclMessageDrivenChannelAdapter` and `KinesisMessageDrivenChannelAdapter`.

I also lowered the default value for the interval from 60 seconds to 5. It's kind of a random value, but 60s seemed too long.

pivotal-issuemaster commented:
@javarno Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=114)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=114) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@javarno Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration-aws&pullRequestId=114)!

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/CheckpointMode.java:
```diff
@@ -39,6 +39,11 @@
+	manual,
+
+	/**
+	 * Checkpoint at fixed time intervals.
```
Need a `@since 2.1` - it is a new public API.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java:
```diff
@@ -229,6 +231,15 @@ public void setCheckpointMode(CheckpointMode checkpointMode) {
+	/**
+	 * Sets the interval between 2 checkpoints. Only used when checkpointMode is periodic.
+	 *
+	 * @param checkpointsInterval interval between 2 checkpoints (in milliseconds)
```
`@since 2.1` and no blank lines in the method JavaDocs.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -355,6 +369,10 @@ private void setAttributesIfNecessary(Object record, Message<?> message) {
 			}
 
+			if (CheckpointMode.manual == KclMessageDrivenChannelAdapter.this.checkpointMode) {
+				messageBuilder.setHeader(AwsHeaders.CHECKPOINTER, checkpointer);
```
I wonder if we can adapt a `RecordProcessorCheckpointer` to the existing `Checkpointer` abstraction in the project...

artembilan replied:
Let's don't forget to add an issue for this!
I really would like to have an aligned model, when some one decides to switch from one Kinesis Channel Adapter to another.

artembilan commented:
@javarno ,

it turns out that we need to release SI-AWS 2.1 version next week, so all the KCL/KPL features have been moved to the next 2.2 version.

The concern about AWS SDK 2 still remains: we definitely need to reconsider to downgreade KCL implementation to SDK 1.
Until all our upstream dependencies are ready for SDK 2.

Thanks for understanding.

javarno commented:
@artembilan : I added the missing `since` annotations and changed the exisitng ones on KPL / KCL related classes to 2.2.0.
The build fails, but so does a fresh clone from master, so I suppose you're already aware of that.

Concerning the `RecordProcessorCheckpointer` abstraction, I think I see what you mean, but could that be done in another issue after the KCL version downgrade ?

I'm still not sure when we'll be able to do the downgrade, our planning is quite full, but we'll do our best.

artembilan commented:
You know I have just pushed the downgrade commit: https://github.com/spring-projects/spring-integration-aws/commit/dd9f3a7925f96471075f5830768ce2020e2c93d6

It would be great if you check it and test for your use-case.

The build should be OK: we had a problem with our repo.
This PR should be rebased to the latest master.

Indeed, we can consider a `RecordProcessorCheckpointer` apply already in the separate issue.

I think after your review and feedback for my change I'll proceed with your changes in the Kinesis Binder to align with what we have here.

Then we are good to proceed with 2.2.0.M1 and all other issues can be fixed later.

Thank you for all the help with effort!

javarno commented:
Nice, just saw your commit.

I did a quick test, and have 2 remarks : 
- when creating the `KinesisClientLibConfiguration`, null is used for the aws credentials provider for kinesis/dynamoDB/cloudwatch -> we should be able to inject the credentials from the binder, either by a setter or in constructor
- in `prepareMessageForRecord`, the array copy is not necessary for KCL 1.9.x (it only fails with 2.x version where the data is read only) -> it can be replaced by `record.getData().array();`

I need to check the checkpoint modes again since the merge modified methods with related code, and  do some more tests (I'll try to do it tomorrow).

artembilan commented:
Well, I don't think that we need credentials since we require external instances for all those clients.
On the binder level we would just create them as beans based on the provided credentials and then inject them to this KCL component.

I don't think that copy of bytes is going to hurt. We are going to come back to AWS SDK v2 eventually anyway, so this one won't be as an obstacle for us any more!

Thank you for you feedback so far!

javarno commented:
I tried using the credentials to iniatlize the beans in the binder, but it doesn't work, at least for the kinesis client, I didn't check the others.
There's a  `KinesisProxy` which is initialized to retrieve the shards list among other things, and it specifically uses  `config.getKinesisCredentialsProvider()` to retrieve the credentials, not the credentials provider inside the kinesis client (see the constructors of `com.amazonaws.services.kinesis.clientlibrary.proxies.KinesisProxy`).

That's why I had to inject the credentials provider in the adapter to make it work. Another possibility would be to inject the whole `KinesisClientLibConfiguration` bean, not sure which is best. If you see another solution, please let me know, I'll try it.

javarno commented:
I merged the master in this branch, the tests I performed are all OK (I had to inject the credentials to do the tests, but removed that code before pushing).

artembilan commented:
@javarno ,

Thank you for your feedback!

So, I did the fix like this: https://github.com/spring-projects/spring-integration-aws/commit/b1cad7e55a105c47182e7f709e967704d7ef6185

Now we alongside with provided AWS clients we also need to inject an external `AWSCredentialsProvider` as it is really required by the internal `KinesisProxy`.

Please, let me know how it works for you and consider to rebase your PR to the latest master with fixing conflicts.
After that I think we are good to go with the final review here and merge.

Then we are good to come back go your Kinesis Binder PR to apply all our effort from here.

artembilan commented:
Please, *rebase* your branch to the `master`. It is impossible to merge this now since there are some conflicts.

javarno commented:
I merged your commit and tested it, it works, thanks !

javarno commented:
Since I already did the merge, but without using rebase, I force pushed it on the branch to avoid doing it again, I hope it's ok.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/CheckpointMode.java:
```diff
@@ -39,6 +39,12 @@
+	 * Checkpoint at fixed time intervals.
+	 * @since 2.2.0
+	 */
+	periodic
```
Your name to the `@author` list.

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java:
```diff
@@ -291,18 +297,36 @@ public void processRecords(List<Record> records, IRecordProcessorCheckpointer ch
+			}
+			else if (KclMessageDrivenChannelAdapter.this.checkpointMode == CheckpointMode.periodic) {
+				// Checkpoint once every checkpoint interval.
+				if (System.currentTimeMillis() > nextCheckpointTimeInMillis) {
```
This `if` can be combined with its wrapping `else if`

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java:
```diff
@@ -229,6 +231,15 @@ public void setCheckpointMode(CheckpointMode checkpointMode) {
+	 * @param checkpointsInterval interval between 2 checkpoints (in milliseconds)
+	 * @since 2.2.0
+	 */
+	public void setCheckpointsInterval(long checkpointsInterval) {
```
Please, your name to the `@author` list

artembilan commented on src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java:
```diff
@@ -1019,6 +1032,13 @@ private void processRecords(List<Record> records) {
 			}
+			else if (CheckpointMode.periodic.equals(KinesisMessageDrivenChannelAdapter.this.checkpointMode)) {
+				// Checkpoint once every checkpoint interval.
+				if (System.currentTimeMillis() > nextCheckpointTimeInMillis) {
```
The same `ifs` combining concern.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration-aws/commit/79ccf55659d50aebad18f82688254d344984f78d after fixing mentioned changes.

Thank you very much for the contribution; looking forward for more!
